src/Lecture2.hs

```diff
@@ -292,4 +416,21 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding e = case extractConstants e of
+  (Just expr, 0) -> expr
+  (Just expr, n) -> Add expr (Lit n)
+  (Nothing, n) -> Lit n
+
+extractConstants :: Expr -> (Maybe Expr, Int)
```

Comment:
You can go one step further and return `([String], Int)` to be more precise about what the function returns. After that, the only thing you need to do is to convert a list of variable names into a nested application of `Add` constructors